Improved AppConfig logic for 1.7.x and 1.8.x (EXPOSUREAPP-3455)

d4rken opened the pull request:
This PR refactores the `AppConfigProvider` to prepare for the changes to the app config in 1.7.x and 1.8.x

* The app config (`ConfigData`) consists now of "per module" sub interfaces, e.g. there is a `RiskCalculationConfig` that is created by the `RiskCalculationConfig.Mapper` such that changes to individual parts of the original protobuf config are hidden and can be worked on in parallel.
* The `AppConfigProvider` now offers a `val appConfig: Flow<ConfigData>` to allow other app components to react to config changes (e.g. reset RiskCalculuation, EXPOSUREAPP-2718 @chris-cwa )
* `ConfigData` exposes time variables (server time and the offset to our local time) to allow other app components to abort e.g. Riskcalclulation if the time is incorrect (EXPOSUREAPP-2998)
* There is a new test menu solely for viewing the current app config, seeing when it was last refreshed and to force a refresh.

## Test
* Smoke test the app
* Review that the tests in the `appconfig` package cover all edge cases you can think off
* Play a bit with the app config test menu

d4rken committed:
Split and hide the protobuf config behind interfaces with individual mappers responsible for creating the desired formats.

d4rken committed:
Merge branch 'release/1.7.x' into feature/3455-more-frequent-riskscore-updates-configs

# Conflicts:
#	Corona-Warn-App/src/deviceForTesters/java/de/rki/coronawarnapp/test/risklevel/ui/TestRiskLevelCalculationFragmentCWAViewModel.kt
#	Corona-Warn-App/src/main/java/de/rki/coronawarnapp/transaction/RiskLevelTransaction.kt
#	Corona-Warn-App/src/test/java/de/rki/coronawarnapp/transaction/RiskLevelTransactionTest.kt

d4rken committed:
Merge branch 'release/1.7.x' into feature/3455-more-frequent-riskscore-updates-configs

# Conflicts:
#	Corona-Warn-App/src/deviceForTesters/java/de/rki/coronawarnapp/test/risklevel/ui/TestRiskLevelCalculationFragmentCWAViewModel.kt
#	Corona-Warn-App/src/main/java/de/rki/coronawarnapp/transaction/RiskLevelTransaction.kt
#	Corona-Warn-App/src/test/java/de/rki/coronawarnapp/transaction/RiskLevelTransactionTest.kt

d4rken committed:
Make the AppConfig observable.
Provide the server time offset.
Offer a lastUpdatedAt timestamp.
Add an app config specific test screen.
Clean up test screens a bit and move debug options out of API test options.

d4rken committed:
Fix test regression due to refactoring (moved code around).

d4rken committed:
Merge branch 'release/1.7.x' into feature/3455-more-frequent-riskscore-updates-configs

d4rken committed:
Store the server timestamp and offset at retrieval.
Switch to config storage via json to be able to store additional meta data fields (i.e. time).

d4rken committed:
KLint and Me have a hate relationship based on both mutual admiration.

d4rken commented:
Can be reviewed, still fixing the last unit tests.

d4rken committed:
Fix time offset parsing being locale dependent.

d4rken committed:
Fix broken unit tests.

d4rken committed:
Improve offset accuracy, move before unzipping.

d4rken committed:
Fix overly long livedata subscription to results (viewmodel scope vs observer scope)

d4rken committed:
Merge branch 'release/1.7.x' into feature/3455-more-frequent-riskscore-updates-configs

d4rken committed:
Add mapping for the new protobuf configs + tests.

d4rken committed:
For cached (retrofit) response, we need to check the cacheResponse and its timestamps
to determine an accurate time offset.

d4rken committed:
Merge branch 'release/1.7.x' into feature/3455-more-frequent-riskscore-updates-configs

d4rken committed:
Exposure a boolean property to tell us when a fallback config is being used.

d4rken committed:
Merge remote-tracking branch 'origin/feature/3455-more-frequent-riskscore-updates-configs' into feature/3455-more-frequent-riskscore-updates-configs

chris-cwa commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/appconfig/download/AppConfigServer.kt:
```diff
@@ -0,0 +1,94 @@
+            if (verificationKeys.hasInvalidSignature(exportBinary, exportSignature)) {
+                throw ApplicationConfigurationCorruptException()
+            }
+            exportBinary!!
```
!! shouldn't be used, change it so that the compiler regards it as non-null

chris-cwa commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/appconfig/mapping/CWAConfigMapper.kt:
```diff
@@ -0,0 +1,43 @@
+
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    internal fun AppConfig.ApplicationConfiguration.getMappedSupportedCountries(): List<String> {
+        if (supportedCountriesList == null) return emptyList()
+
+        if (supportedCountriesList.size == 1 && !VALID_CC.matches(supportedCountriesList.single())) {
+            Timber.w("Invalid country data, clearing. (%s)", supportedCountriesList)
+            return emptyList()
+        }
+
+        return supportedCountriesList
```
convert to when?

d4rken committed:
Hide the observable flow<ConfigData> behind a method that can automatically triggers refreshes.

d4rken committed:
Use a common mapper interface.

ralfgehrer commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/appconfig/ConfigData.kt:
```diff
@@ -0,0 +1,23 @@
+    val updatedAt: Instant
+
+    /**
+     * serverTime + localOffset = updatedAt
```
minor: shouldn't this comment rather go to updatedAt

ralfgehrer commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/appconfig/DefaultConfigData.kt:
```diff
@@ -0,0 +1,14 @@
+    override val localOffset: Duration,
+    override val isFallback: Boolean
+) : ConfigData, ConfigMapping by mappedConfig {
+    override val updatedAt: Instant = serverTime.plus(localOffset)
```
what about 
"* If **[isFallback]** returns true,
     * you should probably ignore the time offset."

ralfgehrer reviewed (commented):
Great work.

d4rken committed:
Address PR comments and KLints.

d4rken committed:
Fix refactoring regression.

d4rken committed:
Improve config unzipping code.

d4rken committed:
Add flag to forward exception thrown during HotDataFlow.kt initialization.

d4rken committed:
Merge branch 'release/1.7.x' into feature/3455-more-frequent-riskscore-updates-configs

d4rken committed:
Don't specify a default context via singleton.

d4rken commented:
Offline mode behavior is not working correctly yet...

d4rken committed:
Merge remote-tracking branch 'origin/feature/3455-more-frequent-riskscore-updates-configs' into feature/3455-more-frequent-riskscore-updates-configs

d4rken committed:
Move download and fallback logic into it's own class just responsible for sourcing the config: "AppConfigSource".
"AppConfigProvider" is now only responsible for making it available.

d4rken committed:
Simplify current concepts for making the app config observable until we have a default configuration.

d4rken committed:
Improve app config test screen, delete options, better feedback.
Show toast instead of crash on errors.

d4rken committed:
Fixed GSON serialization not encoding/decoding the byte array correctly.
Added specific type adapters for instant and duration to get cleaner json.

d4rken committed:
Remove type adapters from base gson due to conflict with CalculationTrackerStorage.

d4rken commented:
Make sure to check offline (airplane mode) behavior too.

ralfgehrer commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/appconfig/AppConfigProvider.kt:
```diff
@@ -1,149 +1,45 @@
-            sanityChecked = sanityChecked.toNewConfig {
-                clearSupportedCountries()
+    suspend fun getAppConfig(): ConfigData = mutex.withLock {
+        Timber.tag(TAG).v("getAppConfig()")
```
Could be removed IMHO

d4rken replied:
I like a bit more verbose logging :blush:

ralfgehrer commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/appconfig/AppConfigSource.kt:
```diff
@@ -0,0 +1,82 @@
+        val (serverBytes, serverError) = try {
+            server.downloadAppConfig() to null
+        } catch (e: Exception) {
+            Timber.tag(TAG).w(e, "Failed to download AppConfig from server .")
```
rather an error (.e.) log?

d4rken replied:
As this would also trigger when no internet is available, it's more of a warning, isn't it? An error will be thrown and logged if then also no fallback config is available.

ralfgehrer commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/appconfig/download/AppConfigServer.kt:
```diff
@@ -0,0 +1,94 @@
+        val offset = Duration(serverTime, localTime)
+        Timber.tag(TAG).v("Time offset was %dms", offset.millis)
+        return ConfigDownload(
+            rawData = rawConfig,
```
above "calculations" could be split into dedicated functions for easier readability

d4rken replied:
Moved the extraction into local extension functions to improve readability.

ralfgehrer reviewed (dismissed):
Successfully tested with and without internet connection on a Pixel (Android 10).

d4rken committed:
We want to default to forced serialization of instant by our converters, instead of using the default serialization which will differ
between Java8.Instant and JodaTime.Instant, to prevent future headaches there, register explicit converters by default,
and overwrite them if necessary (currently only needed for CalculationTrackerStorage.kt).

d4rken committed:
Improve AppConfigServer code readability by moving code into extensions.

d4rken committed:
Merge branch 'release/1.7.x' into feature/3455-more-frequent-riskscore-updates-configs

# Conflicts:
#	Corona-Warn-App/src/main/java/de/rki/coronawarnapp/transaction/SubmitDiagnosisKeysTransaction.kt
#	Corona-Warn-App/src/test/java/de/rki/coronawarnapp/transaction/SubmitDiagnosisKeysTransactionTest.kt

d4rken committed:
Fix merge conflicts

d4rken committed:
Merge branch 'release/1.7.x' into feature/3455-more-frequent-riskscore-updates-configs

harambasicluka reviewed (commented):
tested on device, works like expected 👍

harambasicluka committed:
Merge branch 'release/1.7.x' into feature/3455-more-frequent-riskscore-updates-configs

chris-cwa commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/serialization/adapter/ByteArrayAdapter.kt:
```diff
@@ -0,0 +1,20 @@
+
+    override fun read(reader: JsonReader): ByteArray? = when (reader.peek()) {
+        NULL -> reader.nextNull().let { null }
+        else -> reader.nextString().decodeBase64()!!.toByteArray()
```
```suggestion
        else -> reader.nextString().decodeBase64()?.toByteArray() ?: byteArrayOf()
```

chris-cwa replied:
alternativ:
```suggestion
        else -> reader.nextString().decodeBase64()?.toByteArray()
```

d4rken replied:
This would change the returned data though, if we are not reading a NULL token, then we expect there to be valid data, we want to crash then, not return "null" as if there would be no data.

AFAICT (see [here](https://github.com/square/okio/blob/master/okio/src/commonMain/kotlin/okio/-Base64.kt#L31)), the only way for `decodeBase64` to return null is on invalid data. So if we then return null in the not `NULL` token case, then we treat not existing data and invalid data the same. Would that be desired behavior?

chris-cwa replied:
ok, if a crash is ok with everyone - let's do it!

d4rken replied:
I think the crash needs to happen, otherwise the classes above, using gson (with this adapter) to deserialize something, wouldn't know that there is something wrong with the data if there was no exception. What do you think?

chris-cwa replied:
I believe this app should never crash in production. :( but my opinion is not the only one that counts.

d4rken replied:
That's not what I meant.

The code we are talking about is part of a custom type adapter for data serialization via Gson.

The TypeAdapter does not have enough information to determine whether returning `null` or throwing an exception is the correct approach, so it should not make that decision, but throw an exception so that the "higher level classes" know about this issue and can make the decision as they have more information available.

```kotlin
    fun getLastTimeStampData() : Instant? {
        return try {
            gson.fromJson(someFile)
        } catch (e: Exception) {
            Timber.e(e, "Our data was corrupted, returning EPOCH")
            Instant.EPOCH
        }
    }
```

The `getLastTimeStampData` knows about the data's purpose and can make a decision whether we should return null in an error case, or a fallback value. If we catch the decoding issue in the `ByteArrayAdapter` and just return `null`, then `getLastTimeStampData` could no longer distinguish between "no value exists" and "value exists but data corrupted/invalid".

d4rken replied:
Thinking about it, it would be good to throw a more specific Exception than a Kotlin NPE. Will add that.

d4rken replied:
Let me add a unit test for before it's merged.

harambasicluka committed:
Merge branch 'release/1.7.x' into feature/3455-more-frequent-riskscore-updates-configs

d4rken committed:
Throw a more specific exception if gson type decoding fails due to malformed base64 encoding.

d4rken committed:
Add unit test for Gson ByteArrayAdapter.
